ichnaea/api/views.py

```diff
@@ -94,9 +94,14 @@ def preprocess_request(self):
             except GZIPDecodeError as exc:
                 errors.append({"name": None, "description": repr(exc)})
 
+        if isinstance(request_content, bytes):
+            content = request_content.decode(self.request.charset)
+        else:
+            content = request_content
```

Comment:
This suggests the request content is bytes or something else. What else would it be? Is this a Pyramid thing?

Reply:
For some tests, it was string data. I didn't look too closely, but I can see if there was a pattern.

Reply:
My gut feeling is that this is weird. It should always be either str or bytes and not change. So either Pyramid is weird or something is weird in the tests.

We could leave it like this, but I'm concerned that this weirdness suggests there's a bigger problem somewhere that we probably want to fix. Maybe in a middleware or something.

Reply:
They are all testing the gzip-encoded path, like this:

https://github.com/mozilla/ichnaea/blob/3a7917787cf3c38c0729bbd183aaccdfae6aebe5/ichnaea/api/locate/tests/base.py#L437-L446

By default, the gzip utility function decodes the content with `utf-8`, returning a string:

https://github.com/mozilla/ichnaea/blob/fa9e34694dd436c2b222e4613746333b2fef5119/ichnaea/util.py#L47-L59

I'm guessing this could be improved by passing the request charset to the gzip utility function.

Reply:
I'm not sure what you mean by "passing the request charset to the gzip utility function". How does that ensure that `request_content` is always bytes?

Reply:
Sorry for the confusion. The code in `preprocess_request` just above the changed lines is responsible for conditionally converting `request_content` from bytes to a UTF-8 decoded string, based on the way we call ``util.decode_gzip``:

https://github.com/mozilla/ichnaea/blob/fa9e34694dd436c2b222e4613746333b2fef5119/ichnaea/api/views.py#L89-L95

Because this code writes back to `request_content`, it could either still be untouched bytes or could be an uncompressed, utf-8 decoded string by the time it gets to the changed code.

To ensure that ``util.decode_gzip`` does not try to decode the bytes, we could change the call to:

```
request_content = util.decode_gzip(self.request.body, encoding=None) 
```

Then ``request_content`` would remain bytes, and we'd have a uniform decoding method.

Reply:
Ahhh... got it. My bad--I looked at a bunch of code, but not just above the relevant block in views.py.

Seems like it's better to have it always be bytes, so I like your idea. Having said that, if the various type options exist because of this one section of code, that's a lot less weird than I thought it was. So, whichever way you think is best is fine by me.

Reply:
It think it would be better to tell `decode_gzip` to not try to decode bytes, so that all the byte decoding code is in the same place. I'll make that change in a new commit.

After that change, it looks like the `encoding` parameter could possibly be removed from `decode_gzip` and `encode_gzip`, but that will be a wider change, so new PR.